src/audio/mux/mux_generic.c

```diff
@@ -18,347 +18,503 @@
+			(char *)lookup[0].copy_elem[elem].dest;
+		tmp_frames = size_to_end / frame_bytes;
+		/* check whether there is enough memory for tmp_frames and
+		 * one more channel
+		 */
+		if (size_to_end - (tmp_frames * frame_bytes) >= sample_bytes)
+			tmp_frames++;
+		min_frames = (tmp_frames < min_frames) ? tmp_frames :
+			min_frames;
 	}
 
-	return sample;
+	return min_frames;
 }
 
-/* \brief Demuxing 16 bit streams.
- *
- * Source stream is routed to sink with regard to routing bitmasks from
- * mux_stream_data structure. Each bitmask describes composition of single
- * output channel.
- *
- * \param[in,out] sink Destination buffer.
- * \param[in,out] source Source buffer.
- * \param[in] frames Number of frames to process.
- * \param[in] data Parameters describing channel count and routing.
- */
-static void demux_s16le(struct audio_stream *sink,
-			const struct audio_stream *source, uint32_t frames,
-			struct mux_stream_data *data)
+static uint32_t demux_calc_frames_without_wrap(struct comp_dev *dev,
+					       struct audio_stream *sink,
+					       const struct audio_stream
+						*source,
+					       struct mux_look_up *lookup,
+					       uint32_t frames)
 {
-	int32_t sample;
-	int16_t *dst;
-	uint32_t dst_idx;
-	uint8_t i;
-	uint8_t out_ch;
-
-	for (i = 0; i < frames; i++) {
-		for (out_ch = 0; out_ch < sink->channels; out_ch++) {
-			sample = calc_sample_s16le(source,
-						   i * source->channels,
-						   data->mask[out_ch]);
-
-			/* saturate to 16 bits */
-			dst_idx = i * sink->channels + out_ch;
-			dst = audio_stream_write_frag_s16(sink, dst_idx);
-			*dst = sat_int16(sample);
-		}
+	uint32_t frame_bytes;
+	uint32_t sample_bytes;
+	uint32_t size_to_end; /* size to buffer end */
+	uint32_t min_frames;
+	uint32_t tmp_frames;
+	uint8_t elem;
+
+	comp_dbg(dev, "calc_frames_without_wrap(): frames: %d", frames);
+
+	min_frames = frames;
+
+	for (elem = 0; elem < lookup[0].num_elems; elem++) {
+		/* calculate sources min frames */
+		frame_bytes = audio_stream_frame_bytes(source);
+		sample_bytes = audio_stream_sample_bytes(source);
+		size_to_end = (char *)source->end_addr -
+			(char *)lookup[0].copy_elem[elem].src;
+		tmp_frames = size_to_end / frame_bytes;
+		/* check whether there is enough memory for tmp_frames and
+		 * one more channel
+		 */
+		if (size_to_end - (tmp_frames * frame_bytes) >= sample_bytes)
+			tmp_frames++;
+		min_frames = (tmp_frames < min_frames) ? tmp_frames :
+			min_frames;
+
+		/* calculate sink min frames */
+		frame_bytes = audio_stream_frame_bytes(sink);
+		sample_bytes = audio_stream_sample_bytes(sink);
+		size_to_end = (char *)sink->end_addr -
+			(char *)lookup[0].copy_elem[elem].dest;
+		tmp_frames = size_to_end / frame_bytes;
+		/* check whether there is enough memory for tmp_frames and
+		 * one more channel
+		 */
+		if (size_to_end - (tmp_frames * frame_bytes) >= sample_bytes)
+			tmp_frames++;
+		min_frames = (tmp_frames < min_frames) ? tmp_frames :
+			min_frames;
```

Comment:
I think this is over complex. Ideally we need a stream API that allows us to get avail/free bytes to wrap.
```c

static inline int audio_stream_get_avail_nowrap()
{
    if (buffer->r_ptr > buffer->w_ptr)
        return buffer->end - buffer->r_ptr;
    else
       return buffer->w_ptr - buffer->r_ptr;
} 

demux_calc_frames()
{
    /* get source bytes */
    source_bytes = audio_stream_get_avail_nowrap();

   /* get min sink bytes */
    for (elem) {
        sink_bytes = audio_stream_get_free_nowrap();
         if (sink_bytes < min_sink_bytes)
             min_sink_bytes = sink_bytes;
    }

    /* get min bytes */
    min = source_bytes < min_sink_bytes ? source_bytes : min_sink_bytes;

    /* convert to frames */
    return bytes_to_frames(min);
}
```

Reply:
@lgirdwood 
Ok, I will work on it

Reply:
I've just seen that we have something like: 
```
static inline int
audio_stream_bytes_without_wrap(const struct audio_stream *source,
				const void *ptr)
```
in audio_stream.h

Reply:
That sounds like the one to use. Do you have any new MCPS numbers ?

Reply:
@lgirdwood 
I have modified ```mux/demux_calc_frames_without_wrap``` a little bit. I've added inline ```audio_stream_frames_without_wrap()``` function. I've measured MCPS:
- mux - 48 kHz 8 channels 32/32 with 2 input streams ~8500 MCPS:
```[  2912365.750000] (    13988.833008) c0 mux          4.10      src/audio/pipeline.c:206  perf comp_copy peak plat 508 cpu 8501```

- demux - 48 kHz 8 channels 32/32 with 2 output streams ~9200 MCPS
```[  1159658.916667] (       28.208334) c0 demux        5.15      src/audio/pipeline.c:206  perf comp_copy peak plat 550 cpu 9182```

BTW before merging I will ask our validation for stress tests.

Reply:
@bkokoszx great work ! that's alsmost and order of magnitude faster than before. Ping me when stress tests done and I'll do a final review.